firedrake/slate/slac/optimise.py

```diff
@@ -0,0 +1,55 @@
+from gem.node import MemoizerArg
+from functools import singledispatch
+from itertools import repeat
+import firedrake.slate.slate as sl
+
+
+def optimise(expression):
+    return push_block(expression)
+
+
+def push_block(expression):
+    mapper = MemoizerArg(_push_block)
+    ret = mapper(expression, ())
+    return ret
+
+
+@singledispatch
+def _push_block(expr, self, indices):
+    raise AssertionError("Cannot handle terminal type: %s" % type(expr))
+
+
+@_push_block.register(sl.Transpose)
+def _push_block_transpose(expr, self, indices):
+    return sl.Transpose(*map(self, expr.children, repeat(indices[::-1]))) if indices else expr
+
+
+@_push_block.register(sl.Add)
+@_push_block.register(sl.Negative)
+def _push_block_add(expr, self, indices):
+    return type(expr)(*map(self, expr.children, repeat(indices))) if indices else expr
+
+
+@_push_block.register(sl.Factorization)
+@_push_block.register(sl.Inverse)
+@_push_block.register(sl.Solve)
+@_push_block.register(sl.Mul)
+def _push_block_stop(expr, self, indices):
+    return sl.Block(expr, indices) if indices else expr
+
+
+@_push_block.register(sl.Tensor)
+def _push_block_tensor(expr, self, indices):
+    return sl.Tensor(sl.Block(expr, indices).form) if indices else expr
+
+
+@_push_block.register(sl.AssembledVector)
+def _push_block_assembled_vector(expr, self, indices):
+    return sl.BlockAssembledVector(sl.Block(expr, indices).form, indices) if indices else expr
+
+
+@_push_block.register(sl.Block)
+def _push_block_block(expr, self, indices):
+    indices = expr._indices if not indices else tuple(big[slice(*small)] for small, big in zip(expr._indices, indices))
```

Comment:
Maybe assert that the lengths of `expr._indices` and `indices` match?

Reply:
Is here the right place to do that? I would probably check that in the generation of the Blocks?
Actually I think this should be caught in this assertion here already: https://github.com/firedrakeproject/firedrake/blob/186d644669cabdf2b17118f9ba49a6cf1135d018/firedrake/slate/slate.py#L71

Reply:
I was thinking that if `len(indices) < len(expr._indices)` something might go wrong, but I am not quite sure what.

I think I see what this line does, but I am being stupid. I suppose you're inlining the indexing from the outside?

Reply:
If I understand what you mean correctly, I am inlining from the inside. If I have got the slate expression `A.blocks[:3,:3].blocks[0,0]` I encounter the blocks on the slices first. When I encounter the first `blocks` the `indices` are empty and the slices are in `expr._indices`. When I encounter the second `blocks`, the slices are stored in `indices` and (0,0) in `expr._indices`. So in line 53 I basically say `indices = ((0,1,2)[0], (0,1,2)[0])`

We get `len(indices) < len(expr._indices) ` if we would start off with `A.blocks[:3].blocks[0,0]` This is illegal slate and we already run in an error here https://github.com/firedrakeproject/firedrake/blob/186d644669cabdf2b17118f9ba49a6cf1135d018/firedrake/slate/slate.py#L77 but we should maybe add an assertion to catch this before we fail with a `TypeError: 'NoneType' object cannot be interpreted as an integer`

Reply:
Thanks, makes sense now.

Reply:
Actually I realised  that the fact that this errors is my fault, it should just run through anyways, see PR here https://github.com/firedrakeproject/firedrake/pull/2121.

In any case this won't be an issue for the stuff here, so I will resolve.